### ISSUE TITLE
feature request: Specifying non-differentiable outputs in custom_vjp

### ISSUE THREAD

**Issue by ianwilliamson** [google/jax#3806]
Currently the custom_vjp interface allows one to specify input arguments that should not be differentiated via the `nondiff_argnums` argument. 

It would be useful to have an equivalent option for a function with multiple return values, allowing the user to specify which output arguments should not differentiable. Perhaps it could be named `nondiff_outputs`?

**Comment by mattjj** [comment]
Do you have in mind returning values like arrays with integer dtype, or rather values like functions?

**Comment by ianwilliamson** [comment]
Yes, I was thinking that this would be useful for returning non-JAX types. Functions are one example, but this could also be useful for other non-JAX types (or pytrees of non-JAX types). My understanding is that the input-side option, `nondiff_argnums` is used for this purpose, basically specifying the inputs that JAX should ignore for the purposes of derivative tracing. I was thinking that the proposal here could be the output-side equivalent of that option.

Another use case is having an output that I know JAX should not try to differentiate (either because it's not differentiable or because it would be far too computationally expensive). Perhaps one could argue that there are other ways of getting such quantities out, but sometimes it's just more convenient to return them alongside the differentiable outputs. Thus, it would be useful to tell JAX about these.

**Comment by proteneer** [comment]
We also ran into this recently, and it took us a while to figure out this wasn't supported. Any tips on how to deal with this for the time-being?

**Comment by proteneer** [comment]
not-okay:

``` python
import jax
from jax import custom_vjp
import jax.numpy as jnp

class Aux():
  pass

@custom_vjp
def f(x, y):
  return jnp.sin(x) * y, Aux()

def f_fwd(x, y):
  return f(x, y), (jnp.cos(x), jnp.sin(x), y)

def f_bwd(res, g):
  # we never reach this
  g, _ = g
  cos_x, sin_x, y = res # Gets residuals computed in f_fwd
  return (cos_x * g * y, sin_x * g)

f.defvjp(f_fwd, f_bwd)

f(2.0, 3.0)

grad_fn = jax.grad(f, has_aux=True)
grad_fn(2.0, 3.0)
```

okay:
``` python
import jax
import jax.numpy as jnp

class Aux():
  pass

def f(x, y):
  return jnp.sin(x) * y, Aux()

f(2.0, 3.0)

grad_fn = jax.grad(f, has_aux=True)
grad_fn(2.0, 3.0) # TypeError: <class '__main__.Aux'> is not a valid JAX type

```

**Comment by ianwilliamson** [comment]
What I've done is register the non-differentiable types with JAX as pytrees (associating flatten and unflatten functions). In doing so,  you can put all of the non-differentiable stuff into the metadata (not sure if this is the correct term?) portion of the flattened pytree: https://jax.readthedocs.io/en/latest/jax.tree_util.html#jax.tree_util.register_pytree_node

**Comment by proteneer** [comment]
That's pretty neat - do you have any example code? I'm pretty unfamiliar with the pytree classes

**Comment by ianwilliamson** [comment]
Something along these lines:

```python
import dataclass
import jax.numpy as jnp

@dataclasses.dataclass
class Foo:
   jax_thing: jnp.array
   non_jax_thing: int
   
def flatten(foo):
  return (foo.jax_thing,), (foo.non_jax_thing,)

def unflatten(aux_data, children):
  non_jax_thing, = aux_data
  jax_thing, = children
  return Foo(jax_thing, non_jax_thing)

jax.tree_util.register_pytree_node(Foo, flatten, unflatten)
```

If you had nothing differentiable in `Foo` you can just return an empty tuple as the first output of `flatten()` and ignore the second arg of `unflatten()`.

**Comment by proteneer** [comment]
Sorry for the late reply, and thank you for this. I've verified that it works as intended.
